core/src/main/java/org/infinispan/manager/CacheManagerInfo.java

```diff
@@ -186,6 +197,73 @@ public boolean isTracingEnabled() {
                  .getGlobalConfiguration().tracing().enabled();
    }
 
+   public boolean allCachesReady() {
+      return cacheManager.getCaches().keySet().stream()
+            .allMatch(this::isCacheReady);
+   }
+
+   public boolean isCacheReady(String cacheName) {
+      boolean res = isCacheReadyInternal(cacheName);
+      if (!res) log.debugf("Cache '%s' is not ready", cacheName);
+      return res;
+   }
+
+   private boolean isCacheReadyInternal(String cacheName) {
+      try {
+         GlobalComponentRegistry gcr = SecurityActions.getGlobalComponentRegistry(cacheManager);
+         ComponentRegistry cr = gcr.getNamedComponentRegistry(cacheName);
+
+         // Component registry will be null if the cache is misconfigured or shutdown.
+         // We must retrieve the initialization future to verify the expected status.
+         if (cr == null) {
+            // Unknown cache does not make cluster not ready.
+            CompletableFuture<Cache<?, ?>> cf = cacheManager.getCaches().get(cacheName);
+            if (cf == null) return true;
+
+            if (cf.isDone()) {
+               // A cache was requested to start, but it has failed, e.g., misconfiguration.
+               // We still mark the cluster as ready.
+               if (cf.isCompletedExceptionally())
+                  return isCacheInitializationFailureAllowed(cacheName, cf);
+
+               return cf.join().getStatus().allowInvocations();
+            }
+            // Cache not initialized, which means not ready yet.
+            return false;
+         }
+
+         // Verify if the component registry isn't in shutdown state.
+         if (!cr.getStatus().allowInvocations())
+            return false;
+
+         // Non-clustered caches accepting invocations will be ready.
+         if (!cr.getConfiguration().clustering().cacheMode().isClustered())
+            return true;
+
+         // Clustered caches are ready if they are not rebalacing.
+         DistributionManager dm = cr.getDistributionManager();
+         return dm != null && !dm.isRehashInProgress();
+      } catch (Exception e) {
+         log.tracef(e, "Failed to verify if cache '%s' is ready", cacheName);
+         return false;
+      }
+   }
+
+   private static boolean isCacheInitializationFailureAllowed(String cacheName, CompletableFuture<Cache<?, ?>> cf) {
+      try {
+         return cf
+               .thenApply(CompletableFutures.toTrueFunction())
+               .exceptionally(t -> {
+                  log.tracef(t, "Cache %s not initialized with exception", cacheName);
+                  return CompletableFutures.extractException(t) instanceof CacheConfigurationException;
+               })
+               .join();
+      } catch (Throwable t) {
```

Comment:
Do we still need this catch? We already know the CompletableFuture isDone and isCompletedExceptionally, so I would expect `exceptionally` to always proceed.